### DISCUSSION TITLE
Implementing iDEAL and Giropay payment methods with Multisafe

### DISCUSSION THREAD

**Discussion by sudhirrnanda** [juspay/hyperswitch#4143] [General]
### Context 

Embracing diversity in payments is one of the core tenets of Hyperswitch. And we look forward to integrating a large variety of payment processors (connectors) and payment flows around the world.

Multisafepay is a Dutch fintech company that provides a range of payment solutions and services, including online payment processing, digital wallets, and fraud prevention tools, to businesses throughout Europe. In Hyperswitch parlance, Multisafe is one of the connectors we have integrated with. As of now, hyperswitch supports card payments, apple pay and paypal wallets and klarna (BNPL option) via multisafe

We look forward to including some more critical payment methods through Multisafe

### Problem Statement 

iDeal and Giropay are online banking based payment methods that allow end users to complete payments in their own banking environment.

- iDEAL has an almost 100% reach for Dutch businesses and processes close to 1 Billion transactions annually. 
- Giropay is a major player in the German market, and supports about 32,000 online businesses. 

**The goal is to extend Hyperswitch’s integration with Multisafe to support iDeal and Giropay.**

### How to get started: 

1. [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md) 

2.  [Signup and create a sandbox account on Multisafe](https://www.multisafepay.com/) to get the API Credentials

### How to Implement:

Now you can get started with implementation: 

1. Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch

2. Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3) 

3. Understand the implementation of all different payment flows and [how to add a connector](https://github.com/juspay/hyperswitch/blob/main/add_connector.md) 

4. As mentioned earlier for Multisafe, card payments are implemented. Test out this flow from postman using these [test resources](https://docs.multisafepay.com/docs/testing#credit-and-debit-cards) and understand the API implementation 

5. Study the relevant API Documentation of Multisafe to implement [iDeal](https://docs.multisafepay.com/docs/ideal) and [Giropay](https://docs.multisafepay.com/docs/giropay) through Multisafe

6. Implement iDeal and Giropay using the API references in point 5. Modify the following file to implement iDeal and Giropay

> crates/router/src/connector/multisafepay/transformers.rs
> crates/router/src/connector/multisafepay.rs

7. Test the following flows - Payment, Capture, Void and Refund using postman for your  local implementation and generate relevant screenshots of the test cases.

### Expected Outcomes: 

Submit a **pull request** to the [juspay/hyperswitch](https://github.com/juspay/hyperswitch) repository including the relevant screenshots of the test cases.

_Brownie Points:_ Providing clear and comprehensive documentation and code comments.

Note: We welcome multiple PRs for this problem statement

### More Useful Resources: 

[Setting Up Multisafe on Hyperswitch](https://docs.hyperswitch.io/hyperswitch-cloud/connectors/available-connectors/multisafepay)

**Comment by Toucan4Life** [comment]
Hi! I would like to take this one if no one else is interested :)

**Reply by neerajkumar-n** [reply]
@Toucan4Life - You can start on this and share with the proposal.

**Comment by amribm** [comment]
I would like to work on this!

**Reply by neerajkumar-n** [reply]
@AmeerIbrahimm - Sure, please do work on this and submit your proposal.

Check our hiring wiki to know more: https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Comment by shivankar-p** [comment]
I'd like to give this a shot!

**Reply by neerajkumar-n** [reply]
@shivankar-p - please do

**Comment by aadityaguptaa** [comment]
I would like to work on this project, but I'd like to clarify: Can multiple people share the implementation, or is it limited to one person?

**Reply by neerajkumar-n** [reply]
@aadityaguptaa - multiple people can share the implementation, we would then evaluate. However, please ensure to fill in the google form, so we will be able to reach out to you. 

You can find more details here: https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Comment by pooja-murugiah** [comment]
Heyy! I'm a prefinal year student at NITK, and this seems intriguing to me. 
I'm interested in taking on this challenge.

**Reply by neerajkumar-n** [reply]
@pooja-murugiah - sure, please take a stab at this and submit the details in our form.

Know more here: https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Comment by devnamrits** [comment]
Hi @sudhirrnanda,  @c10n3 , I have created an issue( #4218 )from the above discussion, and along with this, I raised PR #4219 for the same. Please check.

**Reply by neerajkumar-n** [reply]
Sure.

**Comment by shahilsinghyadav** [comment]
Hey, I would like to take this one

**Reply by neerajkumar-n** [reply]
@ssy2306 - Sure, please go ahead and fill in the google form mentioned in our Wiki.

Wiki link: https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Comment by iamfaiz123** [comment]
Please don't close the issue, i would give it a try on the weekend ,i have good experience in rust backend i might be able to fix/add it thanks.

**Reply by neerajkumar-n** [reply]
@iamfaiz123 - sure thing, there are many other hiring challenges which requires rust. You can pick others as well. 

https://github.com/juspay/hyperswitch/discussions/categories/hiring-challenge

**Comment by abn15** [comment]
hii, i would like to work on this issue, can you please tell if there's any deadline to submit my approach, thanks.

**Comment by TakshPanchal** [comment]
## Proposal: **Implementing iDEAL and Giropay payment methods with Multisafe** #4143

### Objective
Add iDEAL and Giropay payment processors in [Multisafepay](https://docs.hyperswitch.io/hyperswitch-cloud/connectors/available-connectors/multisafepay) connector. 
### Background
Currently multisafepay provides various payment processors from Banking Methods, BNPL, Credit and Debit cards, Prepaid cards and Wallets segment. From these hyperswitch's multisafepay connector provides support for credit cards and debit cards, mendate payements, GooglePay, Paypal and Klarna only. iDEAL and Giropay are from "Banking Methods" category. 
#### iDeal
iDeal has direct and re-directed flow which is showed below:
![iDeal payment-flow](https://github.com/juspay/hyperswitch/assets/52617262/6356d5cf-e955-4dad-a7c7-2008b8286bfa)

**direct**:  given **iDEAL issuer code** user customer will be directly redirected to the bank’s 
page
**redirect:** Redirect customer to payment’s page to select their bank.
#### Giropay
giropay has only redirect flow which is exactly same flow as iDEAL’s redirect flow.
![Giropay payment-flow](https://github.com/juspay/hyperswitch/assets/52617262/c4cf915d-ccb4-4062-9595-3a283b56f70b)

### Approach
For iDeal deciding direct or re-direct payment type:
If bank name is provided in `PaymentMethodData` then direct path should be used otherwise redirect.  If direct flow is selected then in `MultisafepayPaymentsRequest` ’s `gateway_info` must have **`issuer_id`** field. 

List of iDEAL issuers and respective code is given below ([source](https://docs.multisafepay.com/reference/listidealissuers))

| Code | Description           |
| ---- | --------------------- |
| 0031 | ABN AMRO              |
| 0761 | ASN Bank              |
| 4371 | bunq                  |
| 1235 | Handelsbanken         |
| 9927 | Nationale Nederlanden |
| 9926 | N26                   |
| 0721 | ING                   |
| 0801 | Knab                  |
| 0021 | Rabobank              |
| 0771 | Regio Bank            |
| 1099 | Revolut               |
| 0751 | SNS Bank              |
| 0511 | Triodos Bank          |
| 0161 | Van Lanschot Bankiers |
| 0806 | Yoursafe              |
Only these banks should be allowed in `PaymentMethodData` for iDEAL.
#### Implementation 

All the changes will be in `connector/multisafepay.rs` file and in the enums and traits for request object

- Add `Ideal` and `Giropay` in `Gateway` enum
- changes in `impl TryFrom<&CheckoutRouterData<&T>> for MultisafepayPaymentsRequest`
    - for `payment_type` : add pattern for redirect “giropay” and “iDeal” and direct “iDeal” payments
        
    - for `gateway` : add `BankRedirectData::Ideal` and `BankRedirectData::Geropay` pattern
        
    - for `gateway_info`: Only for direct “iDeal” payments,
        
        - `issuer_id` : get issuer id from the table above
	    There are also optional `gateway_info` fields for modifying payment behaviours like **`allow_change_amount` , `min_amount` and `max_amount`** which are related to QR code payment in direct flop.

**Reply by TakshPanchal** [reply]
Hey @neerajkumar-n  and team, Thanks for the great opportunity. I am late to the party, but please checkout my proposal. Let me know where I can improve myself. 

I will soon open a PR regarding this Approach.

**Reply by neerajkumar-n** [reply]
Hey @TakshPanchal - haha, no worries. I hope you have filled the google form to review the proposal/PR. If you haven't, recommend you to do that.

**Comment by benodiwal** [comment]
Hi i want to work on this task, can i still pick this one ?

**Comment by sid-rh** [comment]
Hey @neerajkumar-n, I have created an issue #4539 for this and raised a PR #4541.

**Comment by kkendrick-bnpl** [comment]
> ### Context
> Embracing diversity in payments is one of the core tenets of Hyperswitch. And we look forward to integrating a large variety of payment processors (connectors) and payment flows around the world.
> 
> Multisafepay is a Dutch fintech company that provides a range of payment solutions and services, including online payment processing, digital wallets, and fraud prevention tools, to businesses throughout Europe. In Hyperswitch parlance, Multisafe is one of the connectors we have integrated with. As of now, hyperswitch supports card payments, apple pay and paypal wallets and klarna (BNPL option) via multisafe
> 
> We look forward to including some more critical payment methods through Multisafe
> 
> ### Problem Statement
> iDeal and Giropay are online banking based payment methods that allow end users to complete payments in their own banking environment.
> 
> * iDEAL has an almost 100% reach for Dutch businesses and processes close to 1 Billion transactions annually.
> * Giropay is a major player in the German market, and supports about 32,000 online businesses.
> 
> **The goal is to extend Hyperswitch’s integration with Multisafe to support iDeal and Giropay.**
> 
> ### How to get started:
> 1. [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md)
> 2. [Signup and create a sandbox account on Multisafe](https://www.multisafepay.com/) to get the API Credentials
> 
> ### How to Implement:
> Now you can get started with implementation:
> 
> 1. Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch
> 2. Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3)
> 3. Understand the implementation of all different payment flows and [how to add a connector](https://github.com/juspay/hyperswitch/blob/main/add_connector.md)
> 4. As mentioned earlier for Multisafe, card payments are implemented. Test out this flow from postman using these [test resources](https://docs.multisafepay.com/docs/testing#credit-and-debit-cards) and understand the API implementation
> 5. Study the relevant API Documentation of Multisafe to implement [iDeal](https://docs.multisafepay.com/docs/ideal) and [Giropay](https://docs.multisafepay.com/docs/giropay) through Multisafe
> 6. Implement iDeal and Giropay using the API references in point 5. Modify the following file to implement iDeal and Giropay
> 
> > crates/router/src/connector/multisafepay/transformers.rs
> > crates/router/src/connector/multisafepay.rs
> 
> 7. Test the following flows - Payment, Capture, Void and Refund using postman for your  local implementation and generate relevant screenshots of the test cases.
> 
> ### Expected Outcomes:
> Submit a **pull request** to the [juspay/hyperswitch](https://github.com/juspay/hyperswitch) repository including the relevant screenshots of the test cases.
> 
> _Brownie Points:_ Providing clear and comprehensive documentation and code comments.
> 
> Note: We welcome multiple PRs for this problem statement
> 
> ### More Useful Resources:
> [Setting Up Multisafe on Hyperswitch](https://docs.hyperswitch.io/hyperswitch-cloud/connectors/available-connectors/multisafepay)

have you also checked out [Ratio Tech's BNPL for B2B](www.ratiotech.com/boost) options? Curious to learn if you all have seen value like other enterprises implementing new financial strategies with embedded payments.

**Comment by decentral1se** [comment]
Hey, not sure why this is closed. I see https://github.com/juspay/hyperswitch/pull/4541 is on the way, amazing! Is this planned to be merged and as I understand it, this would be the only way to integrate with iDEAL and banks like Triodos with hyperswitch? Thanks for all the work so far.
